### ISSUE TITLE
Allow overriding base URL in CLI

### ISSUE THREAD

**Issue by srid** [srid/emanote#20]
Emanote's default template hardcodes `<base href="/">` via reading the "/" from `index.yaml` (`templates.baseUrl: /`). The default works in live server. But its value differs in certain deployment environments, like GitHub Pages. We should provide a way to override this when generating the static HTML (i.e., outside of live server).

```sh
emanote --override-index-data='{"template": {"baseUrl": "srid.github.io/mynotes"} }' gen
```

Note that this command itself is general, and allows overriding anything in the "index data" (which currently is represented by the top-level `index.yaml`). We can leverage the existing `Data.Aeson.Extra.Merge` to do this.

Goal: continue to use `/` as baseUrl in live server mode, but something else during deployment.

**Comment by hi-learn-triz-and-usit** [comment]
Emanote doesn't currently seem to handle command-line args on its own; it seems to just defer to `ema`'s code: https://github.com/srid/emanote/blob/master/src/Main.hs#L24

Are there any other command line arguments we want? Do you have any thoughts on how you would want `emanote` to handle command line arguments? Should I just use https://github.com/srid/ema/blob/master/src/Ema/CLI.hs as a template?

**Comment by srid** [comment]
@hi-learn-triz-and-usit I was thinking we should be able to write our own top-level optparse-applicative stuff, then compose with `ema`'s `cliParser`  (and `Cli` type).

Something like,

```haskell
data Cli = Cli  { overrideIndexData :: Text,    emaCli :: Ema.Cli  }
```

**Comment by srid** [comment]
Here's an example. Neuron used to do this, by delegating to Rib's parser,

https://github.com/srid/neuron/blob/ce44eeec9ea7298ee3955618c3f838d281f34ec3/src/Neuron/Zettelkasten/CLI.hs

**Comment by srid** [comment]
And then we can change emanote's Main to use [`runEmaWithCli`](https://github.com/srid/ema/blob/56abc9fc16c1a04f8c5b7d84d357cc0c7273edee/src/Ema/App.hs#L67) by passing the resulting CliAction from parser.

**Comment by srid** [comment]
Also, `overrideIndexData` from CLI can be stored in `Model` as well (to be applied in `Meta.getEffectiveRouteMeta`)

**Comment by srid** [comment]
Note to self: once https://github.com/srid/emanote/pull/29 gets merged, we would be able to override just about any file -- not only `index.yaml` but Markdown, HTML templates, etc ... via designating whatever the CLI provides as its own layer (at the very end of stack) in union mounting.

(For the purpose of implementing this particular feature that doesn't matter right now, but it is something to keep in mind).

**Comment by srid** [comment]
> For the purpose of implementing this particular feature that doesn't matter right now,

@hi-learn-triz-and-usit Actually I'm thinking, this feature can be implemented more *easily* using the union mount feature, because you don't have to write any YAML merge code. It also is simpler to reason about from user point of view (just put deploy specific files in a directory!).

---

```
cd ~/mynotebook  # or, `-C ~/mynotebook` argument
emanote --overlays ~/mynotebook/.deploys/github-pages gen
```

The idea here is that emanote is now given optional extra path arguments, that are then passed to union mount as overlays. The overlay argument here specifies the deploy configuration, and this directory would contain:

```
$ cat ~/mynotebook/.deploys/github-pages/index.yaml
template:
  baseUrl: srid.github.io/mynotes
$
```

To implement this, we can use your existing PR and change it to support `--overlays`, then change the following

https://github.com/srid/emanote/blob/f14704cb56da1c8b5741d08ef92f6b5f18a73bb8/src/Emanote/Source/Loc.hs#L7-L16

to use `LocUser (NonEmpty FilePath)`.

or something like that. 

Finally the extra CLI paths can be passed to https://github.com/srid/emanote/blob/f14704cb56da1c8b5741d08ef92f6b5f18a73bb8/src/Emanote/Source/Loc.hs#L18-L19

**Comment by srid** [comment]
```
LocUser (NonEmpty FilePath)
```

Actually it should just be `LocUser FilePath`. However, we break the order assumption ... hmm gotta think.

EDIT: some options available,

- use `LocUser Int FilePath` (where the Ind acts as an index), but that is hackish.
- use https://hackage.haskell.org/package/ordered-containers-0.2.2/docs/Data-Set-Ordered.html

**Comment by srid** [comment]
> `emanote --overlays ~/mynotebook/.deploys/github-pages gen`

This doesn't let the user control the *order* of overlays (we want the deploy overlay to override everything else), so we need to spec a different CLI.

Maybe emanote should dictate the behaviour of `-C` now,

```
# Run live server
emanote 

# Same as above
emanote -C .

# Generate statiuc site
emanote gen

# Same as above
emanote -C . gen

# Generate static sites for GitHub Pages
emanote -C . -C .deploys/github gen

# This is wrong, because the order is incorrect. Right layer overrides the left ones.
emanote -C .deploys/github -C . gen
```

(Perhaps `-C` should now be called `-L` - for "layer", as it is not about 'change dir' anymore)

**Comment by hi-learn-triz-and-usit** [comment]
There's a few things I'm not too fond of with this proposed interface, mostly in terms of potential to confuse users:

* `-C .` is not a no-op, just a default, so sometimes it's correct to omit and other times it's not
* Multiple arguments with order mattering is always easy to get wrong and confusing
* Why complicate things with the command line? The primary purpose is to support non-production deploys that are nevertheless mostly the same

Personally, I think the behavior you're proposing is better suited for a configuration file, and then the only command line choice the user needs to make is selecting one (if the default is not the one they want).

**Comment by srid** [comment]
I agree the CLI spec so far is not satisfactory or optimal.

> The primary purpose is to support non-production deploys that are nevertheless mostly the same

For this issue as originally stated, yes. But we are talking about the more general feature here (we might as well split this discussion and move it to a new issue), for which that is just *one* goal. There can be many (under the same overlay-umbrella). For eg., I want to combine my public notes (from elsewhere) to the current private notebook, and create an unified view. Or to use template themes from a common location, in the current notebook (without having to copy the theme files manually).

> Personally, I think the behavior you're proposing is better suited for a configuration file

That's what the .yaml files are for. And here you are talking about one-level up; the meta configuration that is needed to find these configuration files in the very first place. You are suggesting we add a emanote.dhall (analogous to neuron.dhall), and but this is rather unfitting in the context of overlay design - what if I wanted to override emanote.dhall? Or is it supposed to contain *only* overlay locations? That's overkill for a new configuration file (outside existing configuration system). Just specify it in the CLI and be done with it. It is a unnecessary thing to know for the user.

> Multiple arguments with order mattering is always easy to get wrong and confusing

I don't think it is necessarily confusing. The mount command already has this notion (see below), though it doesn't use multiple options like the above proposal.

```
mount -t overlay overlay -olowerdir=/lower,upperdir=/upper,workdir=/work /merged
```

> -C . is not a no-op, just a default, so sometimes it's correct to omit and other times it's not

Here's another proposal:

```bash
# Does nothing; just prints --help
emanote

# Fails, because requires a path argument
emanote run

# Runs emanote live server on the given folder (the only layer)
emanote run .

# Runs emanote gen command, with the given layers in argument order
emanote gen ./.deploy/github ./

# Run emanote live server on current directory, using common themes and public notes (3 layers)
cd ~/MyPrivateNotes
emanote run ~/myemanotethemes ~/MyPublicNotes ./
```

The difference here is that we *mandate* the path, with no default, so it is less confusing, I think.

**Comment by hi-learn-triz-and-usit** [comment]
> That's what the .yaml files are for. And here you are talking about one-level up; the meta configuration that is needed to find these configuration files in the very first place. You are suggesting we add a emanote.dhall (analogous to neuron.dhall), and but this is rather unfitting in the context of overlay design - what if I wanted to override emanote.dhall? Or is it supposed to contain only overlay locations? That's overkill for a new configuration file (outside existing configuration system). Just specify it in the CLI and be done with it. It is a unnecessary thing to know for the user.

I don't think that adding a new configuration file is any more confusing than adding a command line option. I'm also not sure that adding a configuration file for just one parameter is overkill, nor that it will remain as only one parameter. For example, I could imagine setting parameters in this file and making decisions based on the values of those parameters, such as different headers for the live site vs the demo site.

I think that the two main confusions I have are:
* How do we actually expect the overlay system to be used? Do the different layers know about each other, for example?
* Is there a difference between configuring a site and configuring emanote? What sort of stuff goes in the .yaml files?

**Comment by srid** [comment]
I'm not sure if you appreciate the power of the overlay system.

The overlay system is *already* in use when you run emanote on your notebook (without arguments). It runs emanote with two layers:

1. Layer 1, https://github.com/srid/emanote/tree/master/default
2. Layer 2, current working directory

The default layer provides the following,

- `index.md`: So when you visit the root of your site, it displays a default welcome text.
  - Unless your notebook has a `index.md` of course
- `templates/`: The default HTML templates, so your notebook doesn't have to provide one.
- `index.yaml`: The default settings for the site. [Take a look to see what's in here](https://github.com/srid/emanote/blob/master/default/index.yaml)
  - You can add a `index.yaml` to your notebook, and override only a subset of these settings. Emanote will merge them.
- `favicon.svg`: the only static file (referenced in `index.yaml`)

Now if you add a 3rd layer,

- Layer 3, deploy specific

That solves this issue, because in the 3rd layer you only need to add one file

- `index.yaml` setting only baseUrl

**Comment by srid** [comment]
> Do the different layers know about each other, for example?

No, it behaves like Linux's union mount or overlayfs. The layers are independent of each other. Imagine placing botched mirrors on top of one another. What's the view you get on top?  See https://en.wikipedia.org/wiki/Union_mount

**Comment by hi-learn-triz-and-usit** [comment]
> I'm not sure if you appreciate the power of the overlay system.

I'm sure that I don't. But you know what they say about great power...

I think it's valuable to consider the most likely use case and design the interface around that, even if the backend is astonishingly powerful.

**Comment by srid** [comment]
> I think it's valuable to consider the most likely use case and design the interface around that,

This is exactly how the powerful backend arose (plus, it satisfies other hitherto unimagined uses cases too, as mentioned further above), and in the process I avoided neuron-like code complexity. Good luck implementing these features without the union mount concept; the code will be more complex to deal with (which pain is particularly evident when you come back to it in say one year).

**Comment by srid** [comment]
Compare the code handling the 'default' layer files (templates, index.yaml) before / after https://github.com/srid/emanote/pull/26

**Comment by srid** [comment]
Closing in favour of #35 (but discussion may continue).

**Comment by hi-learn-triz-and-usit** [comment]
Right. I agree that the backend should be powerful. I think that where I may have a different perspective from you is in how much of that power it is valuable and important to expose directly to the end user. I think that new user-facing features should be as simple as possible while addressing the use case; this requires that we recognize simplicity and that we identify the use case, neither of which is a trivial task!

**Comment by srid** [comment]
Lookup volatility-based decomposition from https://rightingsoftware.org - in particular the House building analogy (here, union mount == electricity).

> I think that new user-facing features should be as simple as possible while addressing the use case; 

Try coming up with an alternative design that covers **all** the aforementioned use-cases (and more) that you think is simpler than the layer approach. Arguing is a waste of time right now, at this point (enough has been said).

**Comment by hi-learn-triz-and-usit** [comment]
I don't have a copy of Righting Software handy (it's the first I'm hearing of this book, but I'll definitely look into it). Looking into what I can find online about volatility-based decomposition, it seems to make sense to me. It might be valuable to document this as an explicit part of emanote's design philosophy.

> Try coming up with an alternative design that covers all the aforementioned use-cases (and more) that you think is simpler than the layer approach. 

I think we're talking past each other. I _like_ the layer approach in general. My confusion is that I only see one use case being made explicit in this thread (perhaps due to a failure of reading comprehension on my part):

* When I'm generating static html for deployment I want to override the default base url with a value appropriate for the deployment so that links can be generated correctly.

What are the others that I have missed?

**Comment by srid** [comment]
FWIW, I first came across that book about a week or so ago in FP Slack in `#architecture-and-design` (worth discussing there for more on this topic) -- whilst asking about design in general; in the context of having been burned with neuron's code complexity, vowing not to repeat it in emanote -- and I only read the initial few chapters.

> What are the others that I have missed?

See https://github.com/srid/emanote/issues/20#issuecomment-849043533

> > The primary purpose is to support non-production deploys that are nevertheless mostly the same
> 
> For this issue as originally stated, yes. But we are talking about the more general feature here (we might as well split this discussion and move it to a new issue), for which that is just _one_ goal. There can be many (under the same overlay-umbrella). For eg., I want to combine my public notes (from elsewhere) to the current private notebook, and create an unified view. Or to use template themes from a common location, in the current notebook (without having to copy the theme files manually).

It mentions two more use-cases -- merging public notes with private notebook (both existing in two separate locations, say public noteds on public github repo, but private notes in Dropbox), as well as keeping HTML template themes in a common locations for sharing with multiple notebooks.

The CLI example in https://github.com/srid/emanote/issues/35 uses these use-cases as demonstration too (see the last command).

And a third use-case is mentioned here https://github.com/srid/emanote/issues/20#issuecomment-849216835

> Compare the code handling the 'default' layer files (templates, index.yaml) before / after #26

Basically pulling default config (index.yaml) and HTML templates from emanote's `./defaults` directory, all while allowing the user to override any of them (for eg., they can override just "markdown.tpl" to customize markdown theming), but what the "merge" semantics is not all the same (markdown files are completely replaced, as when you put a index.md; but .yaml files must be merged sanely).

**Comment by srid** [comment]
> Basically pulling default config (index.yaml) and HTML templates from emanote's `./defaults` directory, all while allowing the user to override any of them (for eg., they can override just "markdown.tpl" to customize markdown theming), but what the "merge" semantics is not all the same (markdown files are completely replaced, as when you put a index.md; but .yaml files must be merged sanely).

Not to mention supporting 'hot reload' when any of these files change, including the default location (or in any of the user provided layers, such as the common HTML template themes directory).
